### ISSUE TITLE
Blocklists download/update doesn't work

### ISSUE THREAD

**Issue by ghost** [celzero/rethink-app#952]
The process of downloading/updating blockilists is constantly stuck, the "downloading" circle animation goes on for 1 hour now without any changes.
On 2 different phones.

Could this be fixed?

Is there something wrong going on with the blocklist sources?

----------------

Update 30 minutes later:
- The phone where the update was running got an on-screen message "something went wrong, try again later" and the update stopped.
- On the phone where the download was running, the download just stopped.

Something is certainly wrong.

**Comment by ignoramous** [comment]
Are you from China / Turkmenistan? Some countries block rethinkdns servers, so downloads won't succeed.

If not, consider toggling `Use in-app downloader` (from *DNS* -> *Configure* -> *Advanced*) to see if the blocklist download then works?

If not, and if you know how, can you share the output from `adb logcat` as you attempt a new download (which fails)?

**Comment by ignoramous** [comment]
> Wondering, why it doesn't work with Download Manager? It is not blocked and has internet access

Do you have the settings *Use in-app downloader* turned ON or OFF? If it is turned OFF, then Android's built-in Download Manager is what's working.

**Comment by ignoramous** [comment]
> So, wondering why it doesn't work with Download Manager while it is not blocked and has internet access, it worked before.

My guesses are:
1. Your device is on a custom ROM that has a broken Download Manager.
2. Your device is on latest and greatest Android version we haven't yet tested Rethink on.
3. There's a bug in the way Rethink uses the Download Manager (unlikely).
4. Rethink downloads are blocked either by the ISP or the router or something.

**Comment by ignoramous** [comment]
Graphene and /e/ might have a subtly broken Download Manager... Very likely. 

If you are technical enough, can you share `adb logcat` output as you attempt to download using the built-in Download Manager?

**Comment by hussainmohd-a** [comment]
Looks like we've run into a issue when the VPN is in lockdown mode and the 'Use in-app downloader' is turned off. Essentially, the Android download manager isn't kicking off downloads when this happens.

@A-7666,  To double-check, if you start a download in lockdown mode and then try turning off the lockdown mode, the download manager will kickstart the pending download.

**Comment by iambeingtracked** [comment]
Hey, same issue here

I have a logcat of the problem http://0.vern.cc/E4.png

**Comment by ignoramous** [comment]
> I have a logcat of the problem http://0.vern.cc/E4.png

Thanks. Looks like Rethink on your Android does not have access to read from / write to the Downloads folder? Rethink does request this permission, may be it has been revoked or not granted?

**Comment by iambeingtracked** [comment]
Exactly, it does not request the permission https://files.wherelinux.xyz/all-files-permissions.jpg

**Comment by ignoramous** [comment]
> Exactly, it does not request the permission https://files.wherelinux.xyz/all-files-permissions.jpg

You're right.

Unsure why `WRITE_EXTERNAL_STORAGE` permission was removed in this commit: https://github.com/celzero/rethink-app/commit/a9b1fe45213a1dff84b0e64f35ecf2bd73d9c8de

**Comment by Imold** [comment]
I don't even get the "downloading" circle animation, I tap the "DOWNLOAD BLOCKLISTS" button and nothing happens.

**Comment by npahakis** [comment]
I downloaded Rethinkdns through fdroid on my Galaxy S20 (official ROM but debloated a bit) yesterday and wanted to manage blocklists, but the initial download would not do anything. Resetting the app wouldn't work either. The workaround then was to enable in-app downloader and suddenly it worked. 
I dont recall having had these problems, the last time I hat to install the app in February.

What I noticed was, when I decided to stop the download that was going nothing and subsequently turned off the vpn, the download manager suddenly started downloading the files. This happens every time the blocklist download is started, aborted and the vpn stopped afterwards. But even if the download with deactivated vpn succeeded for all the files, the app couldn't work with that and started the process from scratch.

Why is vpn on even required for the blocklists to download? It seems like that was what prevented the download in the first place.

**Comment by Eollone** [comment]
> 我昨天在我的 Galaxy S20（官方 ROM，但有点膨胀）上通过 fdroid 下载了 Rethinkdns，想要管理阻止列表，但初始下载不会执行任何操作。重置应用程序也不起作用。当时的解决方法是启用应用内下载器，突然间它就起作用了。 我不记得上次在二月份安装该应用程序时遇到过这些问题。
> 
> 我注意到，当我决定停止毫无进展的下载并随后关闭 VPN 时，下载管理器突然开始下载文件。每次阻止列表下载开始、中止以及随后 VPN 停止时都会发生这种情况。但即使使用停用的 VPN 成功下载所有文件，该应用程序也无法使用它并从头开始该过程。
> 
> 为什么黑名单下载还需要打开 VPN？看来这就是阻止下载的首要原因。

I agree, currently downloading blocklist requires Rethink to be running in VPN mode first, which would make it impossible to use another VPN to proxy that download.

**Comment by hussainmohd-a** [comment]
We have made multiple improvements in on-device blocklists downloads and updates in [v0.5.7](https://github.com/celzero/rethink-app/releases/tag/v0.5.7). Hopefully these changes should solve some of the issues reported earlier. 

Please try the latest version and let us know if there are any issues.
